addons/hr_holidays/models/resource.py

```diff
@@ -1,10 +1,50 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from odoo import fields, models, api, _
+from odoo.exceptions import ValidationError
 
 
 class CalendarLeaves(models.Model):
     _inherit = "resource.calendar.leaves"
 
     holiday_id = fields.Many2one("hr.leave", string='Leave Request')
+
+    @api.constrains('date_from', 'date_to', 'calendar_id')
+    def check_compare_dates(self):
+        for record in self:
+            if not record.resource_id:
+                existing_leaves = self.env['resource.calendar.leaves'].search([
+                    '&',
+                    '&',
+                    ('id', '!=', record.id),
+                    ('resource_id', '=', False),
+                    '|',
+                    '|',
+                    '&',
+                    ('date_from', '>', record.date_from),
+                    ('date_from', '<', record.date_to),
+                    '&',
+                    ('date_to', '>', record.date_from),
+                    ('date_to', '<', record.date_to),
+                    '&',
+                    ('date_from', '<=', record.date_from),
+                    ('date_to', '>=', record.date_to),
+                ])
```

Comment:
Can't you do a single `search()` which will take all the leaves (using min(date_from)/max(date_to)) and then filter to see if you have any watching ones?

Here's an example of something similar: https://github.com/odoo/enterprise/blob/master/hr_payroll_holidays/models/hr_leave.py#L26 :-)